### PR TITLE
[Spring] Share application context

### PR DESCRIPTION
**Pull request by dom54**
When executing Cucumber in parallel in combination with `cucumber-spring` users would often encounter a cryptic error message: `No qualifying bean of type 'Steps' available: expected single matching bean but found 1` (#1823). 

This was initially fixed by ensuring that the Spring application context was not shared between threads (#1153, #1148). This however has the  downside that the application context is not shared with JUnit tests either (#1583) and comes at significant performance penalty (#1846). 

The root cause however was a race condition in `DefaultListableBeanFactory.registerBeanDefinition(String beanName, BeanDefinition beanDefinition)`. There is a check to see if the definition already exists. If it does not exist, it replaces the bean definition in a `ConcurrentHashMap` (in a synchronized block as well) but then adds the bean name to a list without checking if it already exists.

By synchronizing inside `SpringFactory.start` we prevent concurrent modification of the application context by `cucumber-spring` This in turn makes it possible to share the application context between threads and with JUnit tests.

Fixes: #1846
Fixes: #1583
Closes: #1848
Closes: #1582

### CODE REVIEW

**Comment by dom54** [comment]
making this the default - share context between all threads

File: spring/src/main/java/io/cucumber/spring/SpringFactory.java [hunk 1]
```diff
@@ -249,8 +251,10 @@ private ConfigurableApplicationContext getContext() {
 
         private void registerGlueCodeScope(ConfigurableApplicationContext context) {
             do {
-                context.getBeanFactory().registerScope(SCOPE_CUCUMBER_GLUE, new GlueCodeScope());
-                context = (ConfigurableApplicationContext) context.getParent();
+                synchronized (context) {
+                    context.getBeanFactory().registerScope(SCOPE_CUCUMBER_GLUE, new GlueCodeScope());
```
Comment (dom54) [inline spring/src/main/java/io/cucumber/spring/SpringFactory.java]: registerScope is NOT threadsafe.  

In AbstractBeanFactory in Spring:

Scope previous = (Scope)this.scopes.put(scopeName, scope); //scopes is a LinkedHashMap

File: spring/src/main/java/io/cucumber/spring/FixBootstrapUtils.java [hunk 2]
```diff
@@ -57,7 +52,7 @@
 			"org.springframework.test.context.web.WebAppConfiguration";
 
     static BootstrapContext createBootstrapContext(Class<?> testClass) {
-        CacheAwareContextLoaderDelegate contextLoader = new DefaultCacheAwareContextLoaderDelegate(contextCache.get());
+        CacheAwareContextLoaderDelegate contextLoader = new DefaultCacheAwareContextLoaderDelegate();
```
Comment (dom54) [inline spring/src/main/java/io/cucumber/spring/FixBootstrapUtils.java]: Not passing in a context will use a static (thread safe) context cache, which will be shared between threads

File: spring/src/main/java/io/cucumber/spring/SpringFactory.java [hunk 3]
```diff
@@ -139,8 +139,10 @@ public void start() {
         notifyContextManagerAboutTestClassStarted();
         if (beanFactory == null || isNewContextCreated()) {
             beanFactory = testContextManager.getBeanFactory();
-            for (Class<?> stepClass : stepClasses) {
-                registerStepClassBeanDefinition(beanFactory, stepClass);
+            synchronized (beanFactory) {
+                for (Class<?> stepClass : stepClasses) {
+                    registerStepClassBeanDefinition(beanFactory, stepClass);
```
Comment (dom54) [inline spring/src/main/java/io/cucumber/spring/SpringFactory.java]: registerStepClassBeanDefinition is not threadsafe. Code in DefaultListableBeanFactory

```java
            if (this.hasBeanCreationStarted()) {
                synchronized(this.beanDefinitionMap) { //its synchronized, but it doesnt check again before adding to the updatedDefinitions list
                    this.beanDefinitionMap.put(beanName, beanDefinition);
                    List<String> updatedDefinitions = new ArrayList(this.beanDefinitionNames.size() + 1);
                    updatedDefinitions.addAll(this.beanDefinitionNames);
                    updatedDefinitions.add(beanName);
                    this.beanDefinitionNames = updatedDefinitions;
```

File: spring/src/main/java/io/cucumber/spring/SpringFactory.java [hunk 4]
```diff
@@ -249,8 +251,10 @@ private ConfigurableApplicationContext getContext() {
 
         private void registerGlueCodeScope(ConfigurableApplicationContext context) {
             do {
-                context.getBeanFactory().registerScope(SCOPE_CUCUMBER_GLUE, new GlueCodeScope());
-                context = (ConfigurableApplicationContext) context.getParent();
+                synchronized (context) {
```
Comment (mpkorstanje) [inline spring/src/main/java/io/cucumber/spring/SpringFactory.java]: I don't think this works. We're swapping out `context` with a different instance.  We should probably use a constant as a monitor.

File: spring/src/main/java/io/cucumber/spring/SpringFactory.java [hunk 5]
```diff
@@ -139,8 +139,10 @@ public void start() {
         notifyContextManagerAboutTestClassStarted();
         if (beanFactory == null || isNewContextCreated()) {
             beanFactory = testContextManager.getBeanFactory();
-            for (Class<?> stepClass : stepClasses) {
-                registerStepClassBeanDefinition(beanFactory, stepClass);
+            synchronized (beanFactory) {
```
Comment (mpkorstanje) [inline spring/src/main/java/io/cucumber/spring/SpringFactory.java]: We also have a synchronization step a few lines up inside `CucumberTestContextManager`. For clarity we could synchronize all operations in the start method.

File: spring/src/main/java/io/cucumber/spring/SpringFactory.java (hunk 4 above)
Comment (dom54) [reply spring/src/main/java/io/cucumber/spring/SpringFactory.java]: the synchronization will be on the first reference to context passed in, but I agree it looks "unclean"

File: spring/src/main/java/io/cucumber/spring/SpringFactory.java (hunk 5 above)
Comment (dom54) [reply spring/src/main/java/io/cucumber/spring/SpringFactory.java]: done

Comment (dom54) [reply spring/src/main/java/io/cucumber/spring/SpringFactory.java]: So now - we synchronize the start method statically. So for all instances of SpringFactory created they will start in serial

File: spring/src/main/java/io/cucumber/spring/SpringFactory.java [hunk 6]
```diff
@@ -26,8 +26,6 @@
 import java.util.Set;
 
 import static io.cucumber.spring.CucumberTestContext.SCOPE_CUCUMBER_GLUE;
-import static io.cucumber.spring.FixBootstrapUtils.createBootstrapContext;
-import static io.cucumber.spring.FixBootstrapUtils.resolveTestContextBootstrapper;
```
Comment (mpkorstanje) [inline spring/src/main/java/io/cucumber/spring/SpringFactory.java]: Removed the `FixBootstrapUtils`. Wasn't needed any more.

File: spring/src/main/java/io/cucumber/spring/SpringFactory.java [hunk 7]
```diff
@@ -129,21 +129,24 @@ private static boolean hasAnnotation(Annotation annotation, Collection<Class<? e
 
     @Override
     public void start() {
-        if (stepClassWithSpringContext != null) {
-            testContextManager = new CucumberTestContextManager(stepClassWithSpringContext);
-        } else {
-            if (beanFactory == null) {
-                beanFactory = createFallbackContext();
+        synchronized (monitor) {
```
Comment (mpkorstanje) [inline spring/src/main/java/io/cucumber/spring/SpringFactory.java]: Replaced the `SpringFactory.class` with `monitor`. The class is accessible by any one. And while I don't expect any one to use it to synchronize on it, there also no need to do so.

Comment (mpkorstanje) [reply spring/src/main/java/io/cucumber/spring/SpringFactory.java]: Also explained the why of it all.

File: spring/src/test/java/io/cucumber/spring/threading/RunParallelCucumberTest.java [hunk 8]
```diff
@@ -1,21 +1,23 @@
...
 import static org.hamcrest.CoreMatchers.is;
 import static org.hamcrest.MatcherAssert.assertThat;
-import static org.junit.jupiter.api.Assertions.assertAll;
 
-class RunParallelCukesTest {
+class RunParallelCucumberTest {
 
-    @Test
-    void test() {
+    @RepeatedTest(100)
```
Comment (mpkorstanje) [inline spring/src/test/java/io/cucumber/spring/threading/RunParallelCucumberTest.java]: We're running 100 repetitions with 100 concurrent scenarios. Should tone this down a bit for CI but good enough.

Comment (mpkorstanje) [reply spring/src/test/java/io/cucumber/spring/threading/RunParallelCucumberTest.java]:
```
[INFO] Cucumber-JVM: Spring ............................... SUCCESS [02:12 min]
```
:rofl:

**Comment by mpkorstanje** [comment]
To keep the history readable PRs are squashed when merging. I've drafted a commit message in your original description of the PR. You should be able to edit it.

**Comment by dom54** [comment]
looks good to me

**Comment by aslakhellesoy** [comment]
Hi @dom54,

Thanks for your making your first contribution to Cucumber, and welcome to the Cucumber committers team! You can now push directly to this repo and all other repos under the cucumber organization! 🍾

In return for this generous offer we hope you will:

* ✅ Continue to use branches and pull requests. When someone on the core team approves a pull request (yours or someone else's), you're welcome to merge it yourself.
* 💚 Commit to setting a good example by following and upholding our [code of conduct](https://cucumber.io/conduct) in your interactions with other collaborators and users.
* 💬 Join the [community Slack channel](http://cucumber.io/support#slack) to meet the rest of the team and make yourself at home.
* ℹ️ Don't feel obliged to help, just do what you can if you have the time and the energy.
* 🙋 Ask if you need anything. We're looking for feedback about how to make the project more welcoming, so please tell us!

On behalf of the Cucumber core team,
Aslak Hellesøy
Creator of Cucumber

**Comment by mpkorstanje** [comment]
Aright then we're done!

**Pull request merged**